File: implementation/scipamato/scipamato-public-web/src/test/java/ch/difty/scipamato/publ/web/paper/browse/SimpleFilterPanelChangeEventTest.java
package ch.difty.scipamato.publ.web.paper.browse;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoMoreInteractions;
import static org.mockito.Mockito.when;

import org.apache.wicket.ajax.AjaxRequestTarget;
import org.apache.wicket.markup.html.form.TextArea;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import nl.jqno.equalsverifier.EqualsVerifier;
import nl.jqno.equalsverifier.Warning;

@RunWith(MockitoJUnitRunner.class)
public class SimpleFilterPanelChangeEventTest {

    private SimpleFilterPanelChangeEvent e;

    @Mock
    private AjaxRequestTarget targetMock, targetMock2;

    @Mock
    private TextArea<?> mockComponent;

    @Before
    public void setUp() {
        when(mockComponent.getId()).thenReturn("id");
        when(mockComponent.getMarkupId()).thenReturn("mId");
    }

    @After
    public void tearDown() {
        verifyNoMoreInteractions(targetMock, targetMock2);
    }

    @Test
    public void canRetrieveTarget() {
        e = new SimpleFilterPanelChangeEvent(targetMock);
        assertThat(e.getTarget()).isEqualTo(targetMock);
    }

    @Test
    public void usingMinimalConstructor_doesNotSetAnySpecialStuff() {
        e = new SimpleFilterPanelChangeEvent(targetMock);
        assertThat(e.getId()).isNull();
        assertThat(e.getMarkupId()).isNull();
    }

    @Test
    public void usingWithId_doesAddId() {
        e = new SimpleFilterPanelChangeEvent(targetMock).withId("foo");
        assertThat(e.getId()).isEqualTo("foo");
        assertThat(e.getMarkupId()).isNull();
    }

    @Test
    public void usingWithMarkupId_doesAddMarkupId() {
        e = new SimpleFilterPanelChangeEvent(targetMock).withMarkupId("bar");
        assertThat(e.getId()).isNull();
        assertThat(e.getMarkupId()).isEqualTo("bar");
    }

    @Test
    public void usingWithIdAndMarkupId_doesAddBoth() {
        e = new SimpleFilterPanelChangeEvent(targetMock).withId("hups")
            .withMarkupId("goo");
        assertThat(e.getId()).isEqualTo("hups");
        assertThat(e.getMarkupId()).isEqualTo("goo");
    }

    @Test
    public void canOverrideTarget() {
        e = new SimpleFilterPanelChangeEvent(targetMock);
        assertThat(e.getTarget()).isEqualTo(targetMock);
        e.setTarget(targetMock2);
        assertThat(e.getTarget()).isEqualTo(targetMock2);
    }

    @Test
    public void consideringAddingToTarget_withIdLessEvent_addsTarget() {
        e = new SimpleFilterPanelChangeEvent(targetMock);
        assertThat(e.getId()).isNull();

        e.considerAddingToTarget(mockComponent);

        verify(targetMock).add(mockComponent);
    }

    @Test
    public void consideringAddingToTarget_withDifferingId_doesNotAddTarget() {
        e = new SimpleFilterPanelChangeEvent(targetMock).withId("otherId")
            .withMarkupId("mId");
        e.considerAddingToTarget(mockComponent);
        verify(targetMock, never()).add(mockComponent);
    }

    @Test
    public void consideringAddingToTarget_withSameIdButNullMarkupId_addsTarget() {
        e = new SimpleFilterPanelChangeEvent(targetMock).withId("id");
        assertThat(e.getMarkupId()).isNull();

        e.considerAddingToTarget(mockComponent);

        verify(targetMock).add(mockComponent);
    }

    @Test
    public void consideringAddingToTarget_withSameIdAndDifferingMarkupId_addsTarget() {
        e = new SimpleFilterPanelChangeEvent(targetMock).withId("id")
            .withMarkupId("otherMarkupId");
        e.considerAddingToTarget(mockComponent);
        verify(targetMock).add(mockComponent);
    }

    @Test
    public void consideringAddingToTarget_withSameIdButSameMarkupId_doesNotAddTarget() {
        e = new SimpleFilterPanelChangeEvent(targetMock).withId("id")
            .withMarkupId("mId");
        e.considerAddingToTarget(mockComponent);
        verify(targetMock, never()).add(mockComponent);
    }

    @Test
    public void equals() {
        EqualsVerifier.forClass(SimpleFilterPanelChangeEvent.class)
            .withRedefinedSuperclass()
            .suppress(Warning.STRICT_INHERITANCE, Warning.NONFINAL_FIELDS)
            .verify();
    }

}

File: public/public-web/src/test/kotlin/ch/difty/scipamato/publ/web/paper/browse/SimpleFilterPanelChangeEventTest.kt
package ch.difty.scipamato.publ.web.paper.browse

import ch.difty.scipamato.common.AjaxRequestTargetSpy
import io.mockk.every
import io.mockk.mockk
import org.amshove.kluent.shouldBeEqualTo
import org.amshove.kluent.shouldBeNull
import org.apache.wicket.markup.html.form.TextArea
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

private const val ID = "id"
private const val MARKUP_ID = "mId"

@Suppress("SpellCheckingInspection")
class SimpleFilterPanelChangeEventTest {

    private val targetSpy = AjaxRequestTargetSpy()
    private val e = SimpleFilterPanelChangeEvent(targetSpy)

    private lateinit var mockComponent: TextArea<String>

    @BeforeEach
    fun setUp() {
        mockComponent = mockk {
            every { id } returns ID
            every { markupId } returns MARKUP_ID
        }
    }

    @AfterEach
    fun tearDown() {
        targetSpy.reset()
    }

    @Test
    fun canRetrieveTarget() {
        e.target shouldBeEqualTo targetSpy
    }

    @Test
    fun usingMinimalConstructor_doesNotSetAnySpecialStuff() {
        e.id.shouldBeNull()
        e.markupId.shouldBeNull()
    }

    @Test
    fun usingWithId_doesAddId() {
        SimpleFilterPanelChangeEvent(target = targetSpy, id = "foo").apply {
            id shouldBeEqualTo "foo"
            markupId.shouldBeNull()
        }
    }

    @Test
    fun usingWithMarkupId_doesAddMarkupId() {
        SimpleFilterPanelChangeEvent(target = targetSpy, markupId = "bar").apply {
            id.shouldBeNull()
            markupId shouldBeEqualTo "bar"
        }
    }

    @Test
    fun usingWithIdAndMarkupId_doesAddBoth() {
        SimpleFilterPanelChangeEvent(target = targetSpy, id = "hups", markupId = "goo").apply {
            id shouldBeEqualTo "hups"
            markupId shouldBeEqualTo "goo"
        }
    }

    @Test
    fun consideringAddingToTarget_withIdLessEvent_addsTarget() {
        e.id.shouldBeNull()
        e.considerAddingToTarget(mockComponent)
        targetSpy.components.size shouldBeEqualTo 1
    }

    @Test
    fun consideringAddingToTarget_withDifferingId_doesNotAddTarget() {
        SimpleFilterPanelChangeEvent(target = targetSpy, id = "otherId", markupId = MARKUP_ID).apply {
            considerAddingToTarget(mockComponent)
        }
        targetSpy.components.size shouldBeEqualTo 0
    }

    @Test
    fun consideringAddingToTarget_withSameIdButNullMarkupId_addsTarget() {
        every { mockComponent.id } returns ID
        every { mockComponent.markupId } returns MARKUP_ID
        SimpleFilterPanelChangeEvent(targetSpy, id = ID).apply {
            markupId.shouldBeNull()
            considerAddingToTarget(mockComponent)
        }
        targetSpy.components.size shouldBeEqualTo 1
    }

    @Test
    fun consideringAddingToTarget_withSameIdAndDifferingMarkupId_addsTarget() {
        every { mockComponent.id } returns ID
        every { mockComponent.markupId } returns MARKUP_ID
        SimpleFilterPanelChangeEvent(target = targetSpy, id = ID, markupId = "otherMarkupId").apply {
            considerAddingToTarget(mockComponent)
        }
        targetSpy.components.size shouldBeEqualTo 1
    }

    @Test
    fun consideringAddingToTarget_withSameIdButSameMarkupId_doesNotAddTarget() {
        SimpleFilterPanelChangeEvent(target = targetSpy, id = ID, markupId = MARKUP_ID).apply {
            considerAddingToTarget(mockComponent)
        }
        targetSpy.components.size shouldBeEqualTo 0
    }
}
